File: Python3/aoc2020/solutions/day_21.py
from __future__ import annotations

from collections import defaultdict
from dataclasses import dataclass
from typing import Dict, Generic, List, Set, TypeVar

from .solution import Solution

T = TypeVar("T")
GraphMatching = Dict[T, T]


class Graph(Generic[T]):
    __slots__ = "g"

    g: Dict[T, List[T]]

    def __init__(self) -> None:
        self.g = defaultdict(list)

    def add_edge(self, u: T, v: T) -> None:
        self.g[u].append(v)
        self.g[v].append(u)

    def find_matching_using_Kuhn_alogrithm(self) -> GraphMatching:
        match: GraphMatching = {}
        visited: Set[T] = set()
        has_augmented = True

        while has_augmented:
            has_augmented = False
            visited.clear()

            for u in self.g:
                if u not in match:
                    has_augmented = has_augmented or self.dfs(
                        u, visited, match)

        return match

    def dfs(self, u: T, visited: Set[T], match: GraphMatching) -> bool:
        if u in visited:
            return False

        visited.add(u)
        for v in self.g[u]:
            if v not in match or self.dfs(match[v], visited, match):
                match[v] = u
                match[u] = v
                return True

        return False


@dataclass
class Food:
    ingredients: List[str]
    allergens: List[str]

    @classmethod
    def from_str(cls, food_text: str) -> Food:
        ingredients_text, allergens_text = food_text.split(" (contains ")
        allergens_text = allergens_text[:-1]

        ingredients = ingredients_text.split()
        allergens = allergens_text.split(", ")

        return cls(ingredients, allergens)


def parse_food_list(foods_text: str) -> List[Food]:
    return [Food.from_str(f) for f in foods_text.splitlines()]


def find_allergen_candidates(food_list: List[Food]) -> Dict[str, Set[str]]:
    all_ingredients = set()
    for food in food_list:
        for ingredient in food.ingredients:
            all_ingredients.add(ingredient)

    allergen_candidates: Dict[str, Set[str]
                              ] = defaultdict(lambda: all_ingredients)
    for food in food_list:
        for allergen in food.allergens:
            allergen_candidates[allergen] = allergen_candidates[allergen] & set(
                food.ingredients)

    return allergen_candidates


def count_allergen_free_ingredients(food_list: List[Food]) -> int:
    allergen_candidates = find_allergen_candidates(food_list)

    possible_allergens = set()
    for allergen_candidate in allergen_candidates.values():
        possible_allergens |= allergen_candidate

    return sum(ing not in possible_allergens for f in food_list for ing in f.ingredients)


def find_allergen_list(food_list: List[Food]) -> str:
    allergen_candidates = find_allergen_candidates(food_list)
    graph = Graph[str]()

    for allergen, candidates in allergen_candidates.items():
        for candidate in candidates:
            graph.add_edge(allergen, candidate)

    matching = graph.find_matching_using_Kuhn_alogrithm()
    foreign_allergens = [matching[a]
                         for a in sorted(allergen_candidates.keys())]

    return ",".join(foreign_allergens)


class Day21(Solution):

    def first_task(self, foods_text: str) -> str:
        food_list = parse_food_list(foods_text)

        return str(count_allergen_free_ingredients(food_list))

    def second_task(self, foods_text: str) -> str:
        food_list = parse_food_list(foods_text)

        return find_allergen_list(food_list)

File: Rust/src/solutions/day_21.rs
use super::solution::{Error, Solution};
use std::cmp::Eq;
use std::collections::{HashMap, HashSet};
use std::hash::Hash;

struct Food<'a> {
    ingredients: Vec<&'a str>,
    allergens: Vec<&'a str>,
}

impl Food<'_> {
    fn new(s: &str) -> Food {
        let mut parts = s.split(" (contains ");
        let ingredients_text = parts.next().unwrap();
        let allergens_text = parts.next().unwrap().trim_end_matches(')');

        Food {
            ingredients: ingredients_text.split_whitespace().collect(),
            allergens: allergens_text.split(", ").collect(),
        }
    }
}

struct Graph<'a, T>
where
    T: Eq + Hash,
{
    g: HashMap<&'a T, Vec<&'a T>>,
}

impl<'a, T> Graph<'a, T>
where
    T: Eq + Hash,
{
    fn new() -> Graph<'a, T> {
        Graph { g: HashMap::new() }
    }

    fn add_edge(&'a mut self, u: &'a T, v: &'a T) -> &mut Graph<'a, T> {
        self.g.entry(u).or_insert_with(Vec::new).push(v);
        self.g.entry(v).or_insert_with(Vec::new).push(u);

        self
    }

    fn find_matching_using_kuhn_algorithm(&self) -> HashMap<&T, &T> {
        let mut matching: HashMap<&T, &T> = HashMap::new();
        let mut visited: HashSet<&T> = HashSet::new();
        let mut has_augmented = true;

        while has_augmented {
            has_augmented = false;
            visited.clear();

            for u in self.g.keys() {
                if !matching.contains_key(u) && !visited.contains(u) {
                    has_augmented |= self.dfs(u, &mut visited, &mut matching);
                }
            }
        }

        matching.iter().map(|(&l, &r)| (l, r)).collect()
    }

    fn dfs(
        &'a self,
        u: &'a T,
        visited: &mut HashSet<&'a T>,
        matching: &mut HashMap<&'a T, &'a T>,
    ) -> bool {
        if visited.contains(u) {
            return false;
        }

        visited.insert(u);

        for v in self.g.get(u).unwrap().iter() {
            if !matching.contains_key(v) || self.dfs(matching.get(v).unwrap(), visited, matching) {
                matching.insert(v, u);
                matching.insert(u, v);
                return true;
            }
        }

        false
    }
}

fn parse_food_list(foods_text: &str) -> Vec<Food> {
    foods_text.lines().map(Food::new).collect()
}

fn find_allergen_candidates<'a>(food_list: &[Food<'a>]) -> HashMap<&'a str, HashSet<&'a str>> {
    let mut allergen_candidates: HashMap<&str, HashSet<&str>> = HashMap::new();
    for food in food_list.iter() {
        let food_ingredients: HashSet<&str> = food.ingredients.iter().copied().collect();

        for allergen in food.allergens.iter() {
            let val = allergen_candidates
                .entry(allergen)
                .or_insert_with(HashSet::new);

            if val.is_empty() {
                *val = food_ingredients.iter().copied().collect();
            } else {
                *val = val.intersection(&food_ingredients).copied().collect();
            }
        }
    }

    allergen_candidates
}

fn count_allergen_free_ingredients(food_list: &[Food]) -> usize {
    let possible_allergens: HashSet<&str> = find_allergen_candidates(food_list)
        .values()
        .fold(HashSet::new(), |acc, el| acc.union(el).copied().collect());

    food_list
        .iter()
        .flat_map(|f| &f.ingredients)
        .filter(|&i| !possible_allergens.contains(i))
        .count()
}

fn find_allergen_list(food_list: &[Food]) -> String {
    let allergen_candidates = find_allergen_candidates(food_list);
    let mut graph = Graph::new();
    let matching = allergen_candidates
        .iter()
        .flat_map(|(allergen, candidates)| candidates.iter().map(move |c| (allergen, c)))
        .fold(&mut graph, |g, (allergen, candidate)| {
            g.add_edge(allergen, candidate)
        })
        .find_matching_using_kuhn_algorithm();

    let mut foreign_allergens: Vec<&str> = allergen_candidates.keys().copied().collect();
    foreign_allergens.sort_unstable();

    foreign_allergens
        .iter()
        .map(|el| **matching.get(el).unwrap())
        .collect::<Vec<&str>>()
        .join(",")
}

pub struct Day21 {}

impl Solution for Day21 {
    fn first_task(&self, foods_text: &str) -> Result<String, Error> {
        let food_list = parse_food_list(&foods_text);

        Ok(count_allergen_free_ingredients(&food_list).to_string())
    }

    fn second_task(&self, foods_text: &str) -> Result<String, Error> {
        let food_list = parse_food_list(&foods_text);

        Ok(find_allergen_list(&food_list))
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_first_task() {
        let test_ingredients_text = "mxmxvkd kfcds sqjhc nhms (contains dairy, fish)
trh fvjkl sbzzf mxmxvkd (contains dairy)
sqjhc fvjkl (contains soy)
sqjhc mxmxvkd sbzzf (contains fish)";

        let test_ingredients = parse_food_list(&test_ingredients_text);

        assert_eq!(count_allergen_free_ingredients(&test_ingredients), 5);
    }

    #[test]
    fn test_second_task() {
        let test_ingredients_text = "mxmxvkd kfcds sqjhc nhms (contains dairy, fish)
trh fvjkl sbzzf mxmxvkd (contains dairy)
sqjhc fvjkl (contains soy)
sqjhc mxmxvkd sbzzf (contains fish)";

        let test_ingredients = parse_food_list(&test_ingredients_text);

        assert_eq!(
            find_allergen_list(&test_ingredients),
            String::from("mxmxvkd,sqjhc,fvjkl")
        );
    }
}
